proposals/fleet-multiple-abm-vpp.md

```diff
@@ -0,0 +1,130 @@
+### Multiple ABM/VPP tokens
+
+- Figma: https://www.figma.com/design/j2M1heOh8eZD6LcUJks6HE/%239956-Add-multiple-Apple-Business-Manager-and-Volume-Purchasing-Program-connections?node-id=2-130&t=3vPRYgndVgzoD8e4-0
+- Issue: https://github.com/fleetdm/fleet/issues/9956
+
+- Constraints/bussines logic:
+    - VPP tokens can be assigned to: "All teams", "No team" or a team. Default is "All teams"
+    - ABM tokens can be assigned to: "No team" or a team. Default is "No team"
+    - Multiple ABM/VPP tokens per team.
+    - Same private key can be used for multiple tokens.
+
+### ABM: gameplan overview
+
+The ADE workflow is summarized here
+
+https://github.com/fleetdm/fleet/blob/0a2a48b6d89ab9e428509eb918724c60c245db60/tools/mdm/apple/glossary-and-protocols.md?plain=1#L216-L235
+
+Currently, Fleet manages a single token, and it does the process described there in a cron job every 30 seconds.
+
+To support multiple tokens, we need to:
+
+- On each cron run, do the process for each token that was uploaded.
+- For each host ingested via ABM, keep track of the token that was used to ingest it.
+- Every time we need to make an ABM call to assign an ADE profile, choose the right token to use for the host.
+
+### VPP: gameplan overview
+
+The VPP token is used to:
+
+- Get a list of available apps for a team
+- Assign an app to a host
+
+To support multiple tokens, we need to:
+
+- Add a form to retrieve all tokens available for a team or "no team" this is `(team OR no team) ∪ all teams`
+- To get a list of available apps: loop over each token and get the available apps
+- To assign an app to a host: loop over each token until we find the first one with available licenses, use it to assign the app to the host.
```

Comment:
> loop over each token until we find the first one with available licenses

Which is at most two tokens, the team's and the "all teams".

Reply:
my current understanding is that can be more (unlimited) tokens per team. Each time the IT admin adds a token, it has the ability to select one or more teams to assign to it.

@georgekarrv was worried about it, and proposed we limit to 1 token to 1 team. The problems with this (and why I proposed we move forward with product's original idea while I was writing the guide) are:

1. Uploading a token, and assigning a token to a team happen in two different stages (two different endpoints are hit.) They wanted to preserve this, so we would need a new database state that's "unassigned", tokens in this state can't be used.
2. We'd need to add UI and back-end validations that are a bit complex (see [this](https://www.figma.com/design/j2M1heOh8eZD6LcUJks6HE/%239956-Add-multiple-Apple-Business-Manager-and-Volume-Purchasing-Program-connections?node-id=5435-9548&t=iQAsbToPV6WFbvtY-4))
3. Noah mentioned that we will probably want to lift this restriction in the near future, and dealing with the migration if you have "unassigned" tokens feels like a pain (or we need to support those forever)

We still are at a good point to reconsider if this raises any alarms.

Reply:
Right right right, sorry, for some reason this didn't stick in my state :) 

> Multiple ABM/VPP tokens per team

That makes sense to me, so it's potentially multiple VPP tokens per team/no team, then also possibly multiple for all teams. Don't know if we should enforce a maximum just to prevent edge cases (maybe in a future release).